c++/src/kj/compat/tls-test.c++

```diff
@@ -818,6 +822,241 @@ KJ_TEST("TLS client certificate verification") {
+                           .then([&](auto stream) {
+                             return test.tlsClient.wrapClient(kj::mv(stream), "example.com");
+                           })
+                           .then([&](auto client) {
+                             return test.writeToServer(kj::mv(client));
+                           });
+
+  auto serverPromise = test.receiver->accept().then([&](auto server) {
+    return test.readFromClient(kj::mv(server));
+  });
+
+  clientPromise.wait(test.io.waitScope);
+  serverPromise.wait(test.io.waitScope);
+}
+
+KJ_TEST("TLS receiver accepts multiple clients") {
+  TlsReceiverTest test;
+  ErrorNexus e;
+
+  test.receiver->run();
+
+  KJ_LOG(INFO, "Requesting a bunch of client connects");
+  constexpr auto kClientCount = 20;
+  auto clientPromises = Vector<Promise<void>>();
+  for (auto i = 0; i < kClientCount; ++i) {
+    auto promise = test.baseReceiver->connect()
+                       .then([&](auto stream) {
+                         return test.tlsClient.wrapClient(kj::mv(stream), "example.com");
+                       })
+                       .then([&](auto client) {
+                         return test.writeToServer(kj::mv(client));
+                       });
+    clientPromises.add(kj::mv(promise));
+  }
+
+  KJ_LOG(INFO, "Requesting and resolving a bunch of server accepts in sequence");
+  for (auto i = 0; i < kClientCount; ++i) {
+    // Resolve each receive in sequence like the Supervisor/Network.
+    test.receiver->accept()
+        .then([&](auto server) {
+          return test.readFromClient(kj::mv(server));
+        })
+        .wait(test.io.waitScope);
+  }
+
+  KJ_LOG(INFO, "Resolving all of our client connects in parallel");
+  joinPromises(clientPromises.releaseAsArray()).wait(test.io.waitScope);
+
+  KJ_LOG(INFO, "Requesting one last server accept that we'll never resolve");
+  auto extraAcceptPromise = test.receiver->accept().then([&](auto server) {
+    return test.readFromClient(kj::mv(server));
+  });
+}
+
+KJ_TEST("TLS receiver does not stall on hung client") {
+  TlsReceiverTest test;
+  ErrorNexus e;
+
+  test.receiver->run();
+
+  constexpr auto kClientCount = 20;
+  auto clientPromises = Vector<Promise<void>>();
+
+  KJ_LOG(INFO, "Requesting the first batch of client connects in parallel");
+  for (auto i = 0; i < kClientCount / 2; ++i) {
+    auto promise = test.baseReceiver->connect()
+                       .then([&](auto stream) {
+                         return test.tlsClient.wrapClient(kj::mv(stream), "example.com");
+                       })
+                       .then([&](auto client) {
+                         return test.writeToServer(kj::mv(client));
+                       });
+    clientPromises.add(kj::mv(promise));
+  }
+
+  KJ_LOG(INFO, "Requesting and resolving a client connect that never does ssl connect");
+  KJ_EXPECT_LOG(ERROR, "Client disconnected during SSL_accept");
+  KJ_ASSERT(test.baseReceiver->connect().wait(test.io.waitScope));
```

Comment:
Perhaps this is overly pedantic, but this doesn't cause a "hung client" as suggested by the test name, it causes a client that connects but then immediately disconnects before the SSL handshake.

This is a great test to include, but it seems worth also adding a case here where we don't just throw away the promise returned from `connect()`, but rather hold onto it to avoid the connection being disconnected.

Reply:
That's a good point. I might also be able to test our timeout mechanism also here.

Reply:
Dang, the AsyncIoProvider is baked in a bit heavily into TlsTest. I could refactor it to be external to the TlsTest to get what I need to make a timer, but it seems a bit heavy for this PR.

Reply:
What do we need a timer for? I was just thinking something like:

```
  ...

  KJ_LOG(INFO, "Requesting and resolving a client connect that never does ssl connect");
  auto hungStream = test.baseReceiver->connect().wait(test.io.waitScope);

  ...
```

And you just leave `hungStream` around until the end of the test, verifying that its presence doesn't get in the way of any other connections getting accepted. The idea is that this would have failed previously (or at least slowed down the test until we gave up on waiting) because we wouldn't have accepted any other connections while waiting on `hungStream`'s handshake.

Reply:
Ah sorry, I hadn't realized you had already done this in your fixup and were just commenting about testing the timeout mechanism being too hard. This looks good 👍